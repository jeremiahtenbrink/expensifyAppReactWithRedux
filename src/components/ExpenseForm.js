import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import { connect } from "react-redux";

export class ExpenseForm extends React.Component {
	constructor ( props )
	{
		super( props );
		
		this.state = {
			description    : props.expense ? props.expense.description : "",
			note           : props.expense ? props.expense.note : "",
			amount         : props.expense ? ( props.expense.amount / 100 ).toString() : "",
			createdAt      : props.expense ? moment( props.expense.createdAt ) : moment(),
			calendarFocused: false,
			errorMessage   : "",
		};
	}
	
	onDescriptionChange = ( e ) => {
		const description = e.target.value;
		this.setState( () => ( { description } ) );
	};
	
	onTextAreaChange = ( e ) => {
		const note = e.target.value;
		this.setState( () => ( { note } ) );
	};
	
	onAmountChange = ( e ) => {
		const amount = e.target.value;
		if ( !amount || amount.match( /^\d{1,}(\.\d{0,2})?$/ ) ){
			this.setState( () => ( { amount } ) );
		}
	};
	
	onDateChange = ( createdAt ) => {
		if ( createdAt ){
			this.setState( () => ( { createdAt } ) );
		}
	};
	
	onFocuseChange = ( { focused } ) => {
		this.setState( () => ( { calendarFocused: focused } ) );
	};
	
	onSubmit = ( e ) => {
		e.preventDefault();
		
		if ( !this.state.description || !this.state.amount ){
			// set error state
			this.setState( () => ( { errorMessage: "Please provide description and amount." } ) );
		} else{
			this.setState( () => ( { errorMessage: "" } ) );
			// send to edit or add expense page
			this.props.onSubmit( {
				description: this.state.description,
				amount     : parseFloat( this.state.amount ) * 100,
				createdAt  : this.state.createdAt.valueOf(),
				note       : this.state.note,
			} );
		}
	};
	
	render ()
	{
		return (
			<div>
				{ this.state.errorMessage && <p>{ this.state.errorMessage }</p> }
				<form onSubmit={ this.onSubmit }>
					<input type="text"
					       placeholder="Description"
					       autoFocus
					       value={ this.state.description }
					       onChange={ this.onDescriptionChange }
					/>
					<input type="text"
					       placeholder="Amount"
					       value={ this.state.amount }
					       onChange={ this.onAmountChange }
					/>
					<SingleDatePicker
						date={ this.state.createdAt }
						onDateChange={ this.onDateChange }
						focused={ this.state.calendarFocused }
						onFocusChange={ this.onFocuseChange }
						numberOfMonths={ 1 }
						isOutsideRange={ () => false }
					/>
					<textarea name="note" id="" cols="20" rows="2"
					          value={ this.state.note }
					          placeholder="Add a note for your expense"
					          onChange={ this.onTextAreaChange }/>
					<button>Add Expense</button>
				
				</form>
			</div>
		);
	}
}

export default connect()( ExpenseForm );
