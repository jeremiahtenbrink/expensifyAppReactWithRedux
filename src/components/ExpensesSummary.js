import React from "react";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import { getExpensesTotal } from "../selectors/expenses-total";
import { connect} from "react-redux";

export const ExpensesSummary = ( { expenseCount, expensesTotal, totalNumberOfExpenses } ) => {
	
	const expenseWord = expenseCount === 1 ? "expense" : "expenses";
	const formattedExpensesTotal = numeral( expensesTotal / 100 ).format( "$0,0.00" );
	
	return (
		<div>
			<h3>
				Showing { expenseCount } of { totalNumberOfExpenses } { expenseWord }:
				
			</h3>
			<h3>
				Totalling: { formattedExpensesTotal }
			</h3>
		</div>
	);
};

const mapStateToProps = ( state ) => {
	const visibleExpenses = selectExpenses( state.expenses, state.filter );
	
	return {
		expenseCount        : visibleExpenses.length === 0 ? 0 : visibleExpenses.length,
		expensesTotal        : getExpensesTotal( visibleExpenses ),
		totalNumberOfExpenses: state.expenses.length,
	};
};

export default connect( mapStateToProps )( ExpensesSummary );


