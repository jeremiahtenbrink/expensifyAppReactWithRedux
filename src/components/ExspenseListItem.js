import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from 'numeral';

const ExspenseListItem = ( { description, amount, createdAt, id } ) => {
	return (
		<div>
			<h3><Link to={ "/edit/" + id }>{ description }</Link></h3>
			<p>{ numeral(amount /100 ).format('$,0.00') }
			--
				{ moment(createdAt).format('MMMM Do, YYYY') }</p>
		
		</div>
	);
};

export default ExspenseListItem;
ExspenseListItem.propTypes = {};
ExspenseListItem.defaultProps = {};

