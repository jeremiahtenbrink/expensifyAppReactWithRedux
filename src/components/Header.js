import { NavLink } from "react-router-dom";
import React from "react";
import { startLogout } from "../actions/auth";
import { connect } from "react-redux";

export const Header = ( {startLogout }) => (
	<header>
		<h1>Expensify</h1>
		<nav>
			<ul>
				<li><NavLink to="/dashboard" activeClassName="is-active" exact={ true }>Dashboard</NavLink></li>
				<li><NavLink to="/create" activeClassName="is-active">Create Expense</NavLink></li>
			</ul>
			<ul>
				<li><button onClick={startLogout}>Logout</button></li>
			</ul>
		</nav>
	</header>
);

const mapDispatchToProps = ( dispatch ) => ( {
	startLogout: () => dispatch(startLogout()),
} );

export default connect( undefined, mapDispatchToProps )(Header);