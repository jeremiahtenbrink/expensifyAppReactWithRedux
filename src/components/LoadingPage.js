import React from 'react';
import PropTypes from 'prop-types';

const LoadingPage = (props) => (

    <div className="loader">
        <img className="loader__image" src="/images/loader.gif" alt=""/>
    </div>
);

LoadingPage.propTypes = {};
LoadingPage.defaultProps = {};

export default LoadingPage;
