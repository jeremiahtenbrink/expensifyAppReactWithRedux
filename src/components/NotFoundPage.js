import { Link } from "react-router-dom";
import React from "react";

const NotFoundPage = () => <div>
	404! I pity the fool who couldn't find me!
	<Link to="/">Go home</Link>
</div>;

export default NotFoundPage;