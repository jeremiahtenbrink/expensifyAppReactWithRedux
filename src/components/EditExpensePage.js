import React, { Component } from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends Component {
	
	onEditExpense = (expense) => {
		this.props.startEditExpense( {
			id: this.props.expense.id,
			updates: expense} );
		this.props.history.push( "/" );
	};
	
	onRemoveExpense = () => {
		
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push( "/" );
	};
	
	render ()
	{
		
		return (
			<div>
				<ExpenseForm id="expense-form"
					expense={ this.props.expense }
					onSubmit={this.onEditExpense}
				/>
				
				<button onClick={this.onRemoveExpense}>Remove
				</button>
			</div>
		
		);
	}
}

const mapStateToProps = ( state, props ) => {
	return {
		expense: state.expenses.find( ( expense ) => expense.id === props.match.params.id ),
	};
};

const mapDispatchToProps = (dispatch) => ({
	startEditExpense: (expense) => dispatch(startEditExpense(expense)),
	startRemoveExpense: (idObject) => dispatch(startRemoveExpense( idObject ))
});

export default connect( mapStateToProps, mapDispatchToProps )( EditExpensePage );