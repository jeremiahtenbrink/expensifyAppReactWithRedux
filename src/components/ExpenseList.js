import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import ExspenseListItem from "./ExspenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = ( props ) => (
	<div>
		{
			props.expenses.length === 0 ? (
				<p>No expenses</p>
			): (
				props.expenses.map( ( expense ) => {
					return <ExspenseListItem key={ expense.id } { ...expense } />;
				} )
			)
		}
		
	</div>
);

const mapStateToProps = state => ( {
	expenses: selectExpenses( state.expenses, state.filter ),
} );

export default connect( mapStateToProps )( ExpenseList );

ExpenseList.propTypes = {};
ExpenseList.defaultProps = {};


