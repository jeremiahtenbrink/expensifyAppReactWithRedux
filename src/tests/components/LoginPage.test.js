
import { shallow } from "enzyme";
import React from "react";
import { LoginPage } from "../../components/LoginPage";

test('should test the rendering of the login page', () => {
	const wrapper = shallow( <LoginPage /> );
	expect( wrapper ).toMatchSnapshot();
	
});

test( "should call start login" +
      " on button click", () => {
	
	const startLogin = jest.fn();
	
	const wrapper = shallow( <LoginPage startLogin={ startLogin }/> );
	
	wrapper.find( "button" ).prop( "onClick" )();
	expect( startLogin ).toHaveBeenCalled();
} );