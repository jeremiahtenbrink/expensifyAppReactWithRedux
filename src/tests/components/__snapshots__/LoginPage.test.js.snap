// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the rendering of the login page 1`] = `
<div>
  <button>
    Login
  </button>
</div>
`;
